### ISSUE TITLE
$CHROOT/sys does not automatically unmount, causes suspend/resume issues

### ISSUE THREAD

**Issue by patgrasso** [dnschneid/crouton#2876]
```
name: jessie
encrypted: no
Entering /media/removable/SD Card/xfce-root/chroots/jessie...
crouton: version 1-20161028111244~master:948ee6dc
release: jessie
architecture: amd64
xmethod: xiwi
targets: xiwi,extension,keyboard,xfce
host: version 8530.96.0 (Official Build) stable-channel gnawty 
kernel: Linux localhost 3.10.18 #1 SMP Tue Oct 4 15:41:50 PDT 2016 x86_64 GNU/Linux
freon: yes
Unmounting /media/removable/SD Card/xfce-root/chroots/jessie...
```

#### Please describe your issue:
So recently I've learned a heck of a lot about the unix filesystem and how linux operates in trying to get crouton to run debian + xfce on my chromebook (acer cb3-131) from an SD card.

One issue I haven't gotten over is that every time I suspend/resume, I would need to either pop the SD card out and back in, or restart the chromebook (if popping it out and back in didn't work). The issue seemed to be that the SD card would not automatically re-mount upon resume.

Why? Apparently it was busy `/media/removable/SD Card: Device or resource is busy`.
`lsof +D /media/removable/SD Card` yielded nothing, which was very puzzling for a while. Then one time, after leaving the chroot, I ran `mount` and found that `/media/removable` had been recursively mounted (and hadn't been unmounted, despite me having exited the chroot) #1419.

So, I commented out [the code that mounts /media](https://github.com/dnschneid/crouton/blob/master/host-bin/enter-chroot#L395). However, this was not the only thing that remained mounted after exiting the chroot.
```
...
tmpfs on /run/crw type tmpfs (rw,nosuid,nodev,noatime,size=128k,mode=700)
/dev/mmcblk1p1 on /media/removable/SD Card type ext4 (rw,nosuid,nodev,noexec,relatime,dirsync,data=ordered)
none on /media/removable/SD Card/xfce-root/chroots/jessie/sys type sysfs (rw,nosuid,nodev,noexec,relatime)
debugfs on /media/removable/SD Card/xfce-root/chroots/jessie/sys/kernel/debug type debugfs (rw,nosuid,nodev,noexec,relatime,gid=605,mode=750)
none on /media/removable/SD Card/xfce-root/chroots/jessie/sys/fs/cgroup type tmpfs (rw,nosuid,nodev,noexec,relatime,mode=755)
cgroup on /media/removable/SD Card/xfce-root/chroots/jessie/sys/fs/cgroup/cpu type cgroup (rw,nosuid,nodev,noexec,relatime,cpu)
cgroup on /media/removable/SD Card/xfce-root/chroots/jessie/sys/fs/cgroup/freezer type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
```
`/run/crw` doesn't matter because it's not on the SD card, so it will persist through suspend/resume. All of the other mounts, however, will break as Chrome OS automatically unmounts the SD card when suspending.

So I tried unmounting them. I thought this would be okay, considering /sys is mounted like so:
```
mount --make-rshared /sys
mount --rbind /sys "$CHROOT/sys"
mount --make-rslave "$CHROOT/sys"
```
However, using `umount` on the mount points above propagated to the bound mount points (the real `/sys`), and when I tried to suspend/resume, Chrome OS couldn't start the UI! So then I did
`sudo mount --make-rslave $CHROOT/sys`
`sudo umount -R $CHROOT/sys`

AND IT WORKED.

**TL;DR:**
So I think the issue occurs [where /sys is mounted manually, as opposed to using bindmount](https://github.com/dnschneid/crouton/blob/master/host-bin/enter-chroot#L534). It seems like everything else is mounted/unmounted to `/run/crouton/...`, except sys remains mounted on the actual SD card.

#### If known, describe the steps to reproduce the issue:
1. From an SD card, use `enter-chroot`
2. Exit the chroot
3. Close the laptop / invoke suspend
4. Resume and navigate to `/media/removable/SD Card`
5. Check to see if it is mounted properly (try to check its contents)
6. If it is empty, try removing `SD Card` -- should get `Device or resource is busy` error

I apologize for the long issue.

**Comment by dnschneid** [comment]
Okay, this might explain why people are still seeing recursive mounts on external media....except I don't understand how running --make-rslave the second time actually fixes things.

I have to admit, I barely understood shared/slave bind mounts enough to implement what's there today; if we're going to fix this, I'd be very open to suggestions.

**Comment by patgrasso** [comment]
I'm not quite sure either. I'll keep digging and debugging to see if I can come up with anything.

https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt

^ This is a pretty helpful resource that details how `mount` and `umount` work for anybody else looking into this.

**Comment by dvwz** [comment]
This seems to be the underlying problem when trying to run a crouton update on external media -- the whole chroot becomes corrupted. A fix would be tremendously appreciated. In the meanwhile, do we have to remove the external media before trying any update?

**Comment by anthem** [comment]
A related problem exists here when attempting to start more than one chroot when both are stored on external media, ultimately due to this line:

https://github.com/dnschneid/crouton/blob/master/host-bin/enter-chroot#L401

Chroot A gets /media mounted to $CHROOT/var/host/media. When chroot B is started, /media is also mounted to B's $CHROOT/var/host/media. Because this path is shared and mounted under chroot A, you end up in a mutually-recursive mount loop, and this ultimately ends up crashing things as the system tries to infinitely mount the subpaths.

A similar situation, termed "mountpoint explosion", as well as many other mount subtree niceties are explained better by this article: https://lwn.net/Articles/690679/

For my immediate purposes, adding "--make-unbindable" to the mount command on line 401 fixes this issue by cutting the mount at $CHROOT/var/host/media from being visible within the SD card on other chroots. I.e., chroot B can't see the mount at B's path /media/removable/SDcard/crouton/A/var/host/media. It only sees an empty directory as it actually exists on the SD card.

It seems there may be other places where this might be needed that I haven't identified. The other locations at least aren't recursive in my case and don't immediately cause crashes, but it's clear that there are still mountpoint explosions going on, as starting one chroot causes the total number of mounts in my kernel to jump from 65 to 167. Starting a second chroot makes it jump from 167 to 678.

**Comment by anthem** [comment]
I believe I've found a better solution for the mount explosion problem when running off sdcard/USB. Rather than just making the /media mount --make-unbindable (which I no longer recommend), you can avoid the whole issue altogether at line 129 of mount-chroot: https://github.com/dnschneid/crouton/blob/master/host-bin/mount-chroot#L129

Change the top-level mount to use --make-slave:

            mount --bind --make-slave "$CHROOTSRC" "$CHROOT"

Now none of the mounts inside chroot A can be seen from chroot B nor do they affect the system mounts. I obtain clean startup/shutdown mount counters (checked with 'mount | wc') when running multiple chroots off an SD card this way.

**Comment by dnschneid** [comment]
@anthem, sorry I missed this; looks like I ended up rediscovering your fix in #3820 

Anyway, you were totally correct :)
